File: Puppet/Puppet/MSAnalyticsResultViewController.m
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 */

#import "Constants.h"
#import "EventLog.h"
#import "MSAnalyticsResultViewController.h"

@interface MSAnalyticsResultViewController ()

@property (weak, nonatomic) IBOutlet UILabel *eventNameLabel;
@property (weak, nonatomic) IBOutlet UILabel *eventPropsLabel;
@property (weak, nonatomic) IBOutlet UILabel *didSentEventLabel;
@property (weak, nonatomic) IBOutlet UILabel *didSendingEventLabel;
@property (weak, nonatomic) IBOutlet UILabel *didFailedToSendEventLabel;

@end

@implementation MSAnalyticsResultViewController

#pragma mark - view controller

-(instancetype)initWithCoder:(NSCoder *)aDecoder {
  self = [super initWithCoder:aDecoder];
  if (self) {
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(willSendEventLog:)
                                                 name:kWillSendEventLog
                                               object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(didSucceedSendingEventLog:)
                                                 name:kDidSucceedSendingEventLog
                                               object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(didFailSendingEventLog:)
                                                 name:kDidFailSendingEventLog
                                               object:nil];
  }
  return self;
}

-(void)dealloc {
  [[NSNotificationCenter defaultCenter] removeObserver:self];
}

#pragma mark - Callbacks

-(void)willSendEventLog:(NSNotification *) notification {
  id log = notification.object;
  self.eventNameLabel.text = [log eventName];
  self.eventPropsLabel.text = [NSString stringWithFormat:@"%d", [log properties].count];
  self.didSendingEventLabel.text = kDidSendingEventText;
  [self reloadCells];
}

-(void)didSucceedSendingEventLog:(NSNotification *) notification {
  id log = notification.object;
  self.eventNameLabel.text = [log eventName];
  self.eventPropsLabel.text = [NSString stringWithFormat:@"%d", [log properties].count];
  self.didSentEventLabel.text = kDidSentEventText;
  [self reloadCells];
}

-(void)didFailSendingEventLog:(NSNotification *) notification {
  id log = notification.object;
  self.eventNameLabel.text = [log eventName];
  self.eventPropsLabel.text = [NSString stringWithFormat:@"%d", [log properties].count];
  self.didFailedToSendEventLabel.text = kDidFailedToSendEventText;
  [self reloadCells];
}

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
  [tableView deselectRowAtIndexPath:indexPath animated:YES];
  switch(indexPath.section) {
    case 1:
      switch (indexPath.row) {
        case 0:
          self.eventNameLabel.text = @"";
          self.eventPropsLabel.text = @"";
          self.didSentEventLabel.text = @"";
          self.didSendingEventLabel.text = @"";
          self.didFailedToSendEventLabel.text = @"";
          break;
        default:
          break;
      }
      break;
    default:
      break;
  }
}

- (void)reloadCells {
  NSMutableArray<NSIndexPath*> *rows = [NSMutableArray new];
  [rows addObject:[NSIndexPath indexPathForRow:0 inSection:0]];
  [rows addObject:[NSIndexPath indexPathForRow:1 inSection:0]];
  [rows addObject:[NSIndexPath indexPathForRow:2 inSection:0]];
  [rows addObject:[NSIndexPath indexPathForRow:3 inSection:0]];
  [rows addObject:[NSIndexPath indexPathForRow:4 inSection:0]];
  [self.refreshControl beginRefreshing];
  [self.tableView reloadRowsAtIndexPaths:rows withRowAnimation:UITableViewRowAnimationNone];
  [self.refreshControl beginRefreshing];
}

@end

File: TestApps/Sasquatch/Sasquatch/ViewControllers/MSAnalyticsResultViewController.swift
import UIKit

class MSAnalyticsResultViewController: UITableViewController {

  // Statistics
  @IBOutlet weak var sendingCountLabel: UILabel!
  @IBOutlet weak var succeededCountLabel: UILabel!
  @IBOutlet weak var failedCountLabel: UILabel!
  
  // Last event
  @IBOutlet weak var eventNameLabel: UILabel!
  @IBOutlet weak var eventIdentifierLabel: UILabel!
  @IBOutlet weak var eventPropertiesCountLabel: UILabel!
  @IBOutlet weak var eventStatusLabel: UILabel!

  @IBAction func onDismissButtonPress(_ sender: Any) {
    self.dismiss(animated: true, completion: nil)
  }
  
  override func viewDidLoad() {
    super.viewDidLoad()
    NotificationCenter.default.addObserver(self, selector: #selector(updateAnalyticsResult),
                                           name: NSNotification.Name.updateAnalyticsResult, object: nil)
  }

  deinit {
    NotificationCenter.default.removeObserver(self)
  }
  
  func updateAnalyticsResult(_ notification: Notification) {
    DispatchQueue.main.async {
      let analyticsResult = notification.object as! MSAnalyticsResult
      self.updateLabels(analyticsResult)
      self.reloadCells()
    }
  }

  func updateLabels(_ analyticsResult: MSAnalyticsResult!) {
    self.sendingCountLabel.text = "\(analyticsResult.sendingEvents.count)"
    self.succeededCountLabel.text = "\(analyticsResult.succeededEvents.count)"
    self.failedCountLabel.text = "\(analyticsResult.sendingEvents.count)"
    
    self.eventNameLabel.text = "\(analyticsResult.lastEvent?.name ?? " ")"
    self.eventIdentifierLabel.text = "\(analyticsResult.lastEvent?.eventId ?? " ")"
    self.eventPropertiesCountLabel.text = "\(analyticsResult.lastEvent?.properties?.count ?? 0)"
    //self.eventStatusLabel.text = "\(analyticsResult.sendingEvents.count)"
  }
  
  func reloadCells() {
    var rows = [IndexPath]()
    let rowsInSection = Int(tableView.numberOfRows(inSection: 0))
    for row in 0..<rowsInSection {
      rows.append(IndexPath(row: row, section: 0))
    }
    tableView.reloadRows(at: rows, with: .none)
  }
}

